### PR TITLE
ComplexMedia, -filter_complex and -map features

### LINKED ISSUE #386: Can I create multiple custom filters on an image then overlay to the video
Is it possible to add multiple custom filters on an image and overlay it to the video? In the command below it has to refer to the input. If it possible to add filters to the image first, then I still need to add all those filters to video eventually. Is there a solution for this?

Commands I will be using: rotate, scale, fade, drawbox, drawtext. Literally making things fly over a video D:

Command example:
ffmpeg -i video.mp4 -loop 1 -i image.jpg -filter_complex \ "[1]scale=-1:600[i0]; \ [i0]drawbox=color=white:t=15[i1]; \ [i1]format=yuva420p,fade=in:st=0.4:d=0.2:alpha=1,fade=t=out:st=4.1:d=0.2:alpha=1[i2]; \ [i2]rotate='0.1*sin(20*PI/50*t)':ow=rotw(iw):oh=roth(ih):c=black@0[i3]; \ [0][i3]overlay=x='if(lte(-w+(t-0.3)*2500,0),-w+(t-0.3)*2500,0)':y=-60:shortest=1" \ -c:v libx264 -vb 10M output.mp4

### LINKED ISSUE #410: Can i create a slideshow video from images ?
Can i create a slideshow video from images by PHP-FFMpeg?

like below command:
```
ffmpeg -framerate 20 \
-loop 1 -t 0.5 -i 1.jpg \
-loop 1 -t 0.5 -i 2.jpg \
-loop 1 -t 0.5 -i 3.jpg \
-loop 1 -t 0.5 -i 4.jpg \
-c:v libx264 \
-filter_complex " \
[1:v][0:v]blend=all_expr='A*(if(gte(T,0.5),1,T/0.5))+B*(1-(if(gte(T,0.5),1,T/0.5)))'[b1v]; \
[2:v][1:v]blend=all_expr='A*(if(gte(T,0.5),1,T/0.5))+B*(1-(if(gte(T,0.5),1,T/0.5)))'[b2v]; \
[3:v][2:v]blend=all_expr='A*(if(gte(T,0.5),1,T/0.5))+B*(1-(if(gte(T,0.5),1,T/0.5)))'[b3v]; \
[0:v][b1v][1:v][b2v][2:v][b3v][3:v]concat=n=7:v=1:a=0,format=yuv420p[v]" -map "[v]" out.mp4
```

### LINKED ISSUE #461: Palettegen?
How would I do palettegen with this?  

The 2 commands I use are..

Generating the palette: 
`ffmpeg -y -ss 7 -i clip.mp4 -vf fps=30,scale=480:-1:flags=lanczos,palettegen palette.png`

Generating the gif from palette
`ffmpeg -ss 7 -i clip.mp4 -i palette.png -filter_complex "fps=30,scale=480:-1:flags=lanczos[x];[x][1:v]paletteuse" clip.gif`

### LINKED ISSUE #477: FFMPEG Video Audio Masking
Hello FFMpeg writer

It look really great but I am here now going to mask a voice, that must be get from some text-to-speech API. To achieve this, what i am doing is

1. Prepare a text-to-speech audio by the logged in user's name
Any text-to-speech api. for now i am using ; `http://duncan3dc.github.io/speaker/`

2. Prepare and Audio (abc.mp3) that will be merged/mapped into the resulted video
`ffmpeg -i *mainVideo.mp4* -i *textToSpeechGetUserName.mp3* -filter_complex "[0:a]atrim=0:1,asetpts=PTS-STARTPTS[aud1]; [1:a]atrim=0:1,asetpts=PTS-STARTPTS[aud2]; [0:a]atrim=start=1,asetpts=PTS-STARTPTS[aud3]; [aud1][aud2][aud3]concat=n=3:v=0:a=1[aout]" -map "[aout]" -c:v copy *media/movie/result.mp3* 2>&1`

3. Create a video masked with the newly built audio
`ffmpeg -i *mainVideo.mp4* -i *media/movie/result.mp3* -map 0:v -map 1:a -c:v copy *media/result.mp4* 2>&1`

I want to make it using this library as putting the shell command into the code is not a good practice. 

OR I appreciate if you could suggest any other better and optimized solution.

Thanks a LOT!

### LINKED ISSUE #590: Overlaying Audio Files.
| Q              | A
| -------------- | ---
| Bug?           | no
| New Feature?   | Yes
| Version Used   | Specific tag or commit sha
| FFmpeg Version | FFmpeg or AVConv and version
| OS             | Your OS and version

I'd like the request to be able to apply complex filters with delays as seen
[here](https://superuser.com/questions/1172109/how-to-concatenate-several-short-audio-mp3-files-with-a-specific-overlap-duratio?noredirect=1&lq=1), I didn't see anything remotely similar documented, and I was wonder if there was an option to do a "raw" command. 

It would be exactly this command. 

ffmpeg -i 1.mp3 -i 2.mp3 -i 3.mp3 -i 4.mp3
   -filter_complex
     "[1]adelay=delay1|delay1[b];
      [2]adelay=delay2|delay2[c];
      [3]adelay=delay3|delay3[d];
      [0][b][c][d]amix=4"
merged.mp3

In order to overlay several audio files at differing points of time.

### LINKED ISSUE #640: How to add transparent watermark ?
| Q              | A
| -------------- | ---
| Bug?           | no
| New Feature?   | maybe
| Version Used   | ^0.13.0
| FFmpeg Version | ^0.13.0
| OS             | Your OS and version

#### Actual Behavior

I want to add transparent watermark using following code. 
`$video->filters()->custom('movie=logo.png [watermark]; [in][watermark] overlay=main_w-overlay_w-5:5 [out]');`

It applies watermark but without transparency.

couldn't find any documentation as well.    

#### Expected Behavior

Video with transparent overlay, watermark

#### Possible Solutions

If we do it with -filter_complex, we can do it, but i can't able to apply -filter_complex using filters()->custom() function.

### PR DESCRIPTION
**Pull request by CaliforniaMountainSnake**
| Q                  | A
| ------------------ | ---
| Bug fix?           | no
| New feature?       | yes
| BC breaks?         | no
| Deprecations?      | no
| Related issues/PRs | fixes #386, fixes #410, fixes #461, fixes #477, fixes #590, fixes #640
| License            | MIT

#### What's in this PR?

I found all issues with key "filter_complex" and researched all code base and I didn't find any possibilities to use `-filter_complex` option, except of run raw commands.
So, I decided to develop this feature myself.

#### Why?

This PR adds the possibilities to create commands with multiple inputs and multiple outputs, to use complex filtergraph instead of simple filtergraph, to do complicated processing of input files inside one ffmpeg command using `-filter_complex` option.

#### Example Usage
This PR adds the ComplexMedia type, that allows you to run complicated ffmpeg commands like this:
```php
$inputs = array(
    'video_1.mp4',
    'video_2.mp4',
    'video_3.mp4',
    'video_4.mp4',
);

$complexMedia = $ffmpeg->openComplex($inputs);
$complexMedia->filters()
    ->custom('[0:v]', 'negate', '[v0negate]')
    ->custom('[1:v]', 'edgedetect', '[v1edgedetect]')
    ->custom('[2:v]', 'hflip', '[v2hflip]')
    ->custom('[3:v]', 'vflip', '[v3vflip]')
    ->xStack('[v0negate][v1edgedetect][v2hflip][v3vflip]', XStackFilter::LAYOUT_2X2, 4, '[resultv]');
$complexMedia
    ->map(array('0:a'), new Mp3(), 'video_1.mp3')
    ->map(array('1:a'), new Flac(), 'video_2.flac')
    ->map(array('2:a'), new Wav(), 'video_3.wav')
    ->map(array('3:a'), new Aac(), 'video_4.aac')
    ->map(array('[resultv]'), new X264('aac', 'libx264'), 'output.mp4')
    ->save();
```

#### BC Breaks/Deprecations

ComplexMedia extends from deep AbstractMediaType class and uses its own commands processing, like Concat class.
And doesn't change other existed classes.
I also created unit and functional tests of the ComplexMedia.

### CODE REVIEW

**Comment by CaliforniaMountainSnake** [comment]
I have added checking of the ffmpeg version in ComplexMediaTest::testXStackFilter().
So, with #699 all test will be passed)

**Comment by CaliforniaMountainSnake** [comment]
@Romain, @jens1o, what do you think about this PR?

File: README.md [hunk 1]
```diff
@@ -520,6 +520,82 @@ $video
...
+This class has been developed primarily to use with `-filter_complex`.
+
+So, its `filters()` method accepts only filters that can be used inside `-filter_complex` command.
+ComplexMedia already contains some built-in filters.
+
+#### Base usage
+For example:
+
+```php
+$complexMedia = $ffmpeg->openComplex(array ('video_1.mp4', 'video_2.mp4'));
```
Comment (jens1o) [inline README.md]: `$complexMedia = $ffmpeg->openComplex(array('video_1.mp4', 'video_2.mp4'));`

File: README.md [hunk 2]
```diff
@@ -520,6 +520,82 @@ $video
...
+This code takes 4 input videos, then the negates the first video, stores result in `[v0negate]` stream, detects edges in the second video, stores result in `[v1edgedetect]` stream, horizontally flips the third video, stores result in `[v2hflip]` stream, vertically flips the fourth video, stores result in `[v3vflip]` stream, then takes this 4 generated streams ans combine them in one 2x2 collage video.
+Then saves audios from the original videos into the 4 different formats and saves the generated collage video into the separate file.
+
+As you can see, you can take multiple input sources, perform the complicated processing for them and produce multiple output files in the same time, in the one ffmpeg command.
+
+#### Just give me map!
+You do not have to use `-filter_complex`. For example, just extract the audio from the video:
+
+```php
+$complexMedia = $ffmpeg->openComplex(array ('video.mp4'));
```
Comment (jens1o) [inline README.md]: `$complexMedia = $ffmpeg->openComplex(array('video_1.mp4'));`

**Comment by jens1o** [comment]
I like the general idea of this and it seems quite well engineered. However, I'd like to see a (rather cheap) runtime check whether the FFMpeg binary version is >= 4.1 (before anyone complains why this great new feature just doesn't work). :)

I'd like to see @alexander-schranz's opinion as well.

**Comment by CaliforniaMountainSnake** [comment]
OK, I think this is a good idea. The overhead to run "ffmpeg -version" is negligible relative to heavy video transcoding operations. I have updated the code.

**Comment by alexander-schranz** [comment]
@jens1o It looks very complex and for me its to connected to the complex filter.
If we implement this we should check what have really todo with the complex filter and what has just todo with opening multiple medias. So for example I think `MediaCollection` would be a better name then `ComplexMedia` for holding an multiple Medias.
I think it would be first better to split the pull request a little bit up to see whats first needed to have multiple inputs and outputs. But thats just my personal opinion.

**Comment by CaliforniaMountainSnake** [comment]
I don't think that `MediaCollection` is a better name for this.
It is not a _collection_ in a programming sense. It doesn't contain any objects, over which you can perform some operations. It is not a data structure. It is just a wrapper to build and to run specific ffmpeg command (like anything else `Media` class).

>what have really todo with the complex filter and what has just todo with opening multiple medias or outputing multiple medias

You always work with a _complex filtergraph_ when you have multiple inputs or outputs.
You can to do it implicitly when you only set `-map` options (not set `-filter_complex` explicitly) and have multiple outputs, but it is still a complex filtergraph.

File: src/FFMpeg/Media/ComplexMedia.php [hunk 3]
```diff
@@ -0,0 +1,424 @@
...
+                . '; your ffmpeg version is ' . $currentVersion);
+        }
+    }
+
+    /**
+     * @return array
+     */
+    protected function buildCommand()
+    {
+        $globalOptions = array('threads', 'filter_threads', 'filter_complex_threads');
```
Comment (jens1o) [inline src/FFMpeg/Media/ComplexMedia.php]: Could you please document the newly created `filter_threads` and `filter_complex_threads` as well? :)

**Comment by jens1o** [comment]
@alexander-schranz Thanks for taking a look! :)

However, I don't really understand what you're suggesting. When handling multiple media inputs together (for example doing a splitscreen), you would have to work with these FFMpeg filter methods. Could you please elaborate?

File: src/FFMpeg/Media/ComplexMedia.php (hunk 3 above)
Comment (CaliforniaMountainSnake) [reply src/FFMpeg/Media/ComplexMedia.php]: Hmm, those options are documented in the official man of the ffmpeg: https://ffmpeg.org/ffmpeg.html#Advanced-options

I just added the possibility of the configuration them for user.

Do you want me to add this information into the readme?

Comment (alexander-schranz) [reply src/FFMpeg/Media/ComplexMedia.php]: @CaliforniaMountainSnake would then not be `AdvancedMedia` be a better name when they are documented as advanced option in ffmpeg? /cc @jens1o

**Comment by alexander-schranz** [comment]
@jens1o I just think the naming `ComplexMedia` is to connected to the `filter_complex` parameter and I think thats not the only use case for this type of Media or do all this options only work with `filter_complex` media call then its ok for me calling it `ComplexMedia`, I'm sadly not familiar with this ffmpeg options.
If I read @CaliforniaMountainSnake [comment](https://github.com/PHP-FFMpeg/PHP-FFMpeg/pull/700#discussion_r387233643) correctly they are documented as *advanced options*  so maybe `AdvancedMedia` would be then a better name or? Correct me if I did misinterpret this.

**Comment by CaliforniaMountainSnake** [comment]
@alexander-schranz, maybe you are right. Developed class can work not only with `filter_complex` (but this was a primarily target).
For example, it can work with `-map` options without setting the `filter_complex` option.

If it is an only remark before merge, I'll rename the class to `AdvancedMedia `.

**Comment by CaliforniaMountainSnake** [comment]
Do you have any other remarks?

File: src/FFMpeg/Driver/FFMpegDriver.php [hunk 4]
```diff
@@ -54,4 +55,20 @@ public static function create(LoggerInterface $logger = null, $configuration = a
...
+     * Get ffmpeg version.
+     *
+     * @return string
+     * @throws RuntimeException
+     */
+    public function getVersion()
+    {
+        preg_match('#version\s(\S+)#', $this->command('-version'), $version);
+        if (!isset($version[1])) {
+            throw  new RuntimeException('Cannot to parse the ffmpeg version!');
```
Comment (jens1o) [inline src/FFMpeg/Driver/FFMpegDriver.php]:
```suggestion
            throw new RuntimeException('Cannot to parse the ffmpeg version!');
```

File: src/FFMpeg/Media/ComplexMedia.php (hunk 3 above)
Comment (jens1o) [reply src/FFMpeg/Media/ComplexMedia.php]: I agree that the naming scheme should be as consistent as possible.

**Comment by alexander-schranz** [comment]
@CaliforniaMountainSnake @jens1o I would definitely go then with `AdvancedMedia` and `openAdvanced`. To get ride of the connection to complex filter as it doesn't have always todo with that.

**Comment by CaliforniaMountainSnake** [comment]
@alexander-schranz, @jens1o, ok, I applied your remarks)

**Comment by jens1o** [comment]
Thank you very much. I'll test this manually before the next release but it seems pretty good!

Thanks again, also with tackling our comments and concerncs. :)

**Comment by CaliforniaMountainSnake** [comment]
@jens1o, Hello man, what about new release? I just need a git tag for this)

**Comment by jens1o** [comment]
@CaliforniaMountainSnake Sure, thanks for the ping! Released now.

**Pull request merged**